director/cache_ads.go

```diff
@@ -442,15 +442,25 @@ func updateLatLong(ad *server_structs.ServerAd) error {
 	return nil
 }
 
-// Apply downtimes provided by Origin/Cache servers to the director's in-memory state for the given server.
+// Apply downtimes provided by Origin/Cache servers (<= v7.21) to the director's in-memory state for the given server.
 // This function treats a nil or empty slice as "no downtimes", clearing any stale
 // serverFiltered entry and removing cached server downtimes.
 func applyServerDowntimes(serverName string, downtimes []server_structs.Downtime) {
+	// If the server is >= v7.22, it sets the serverID in the downtime entry.
+	// Its downtimes are already handled by the Registry polling (updateDowntimeFromRegistry),
+	// so we don't need to do anything here
+	for _, dt := range downtimes {
+		if dt.ServerID != "" {
+			log.Debugf("The downtimes set by the admin of Server %q have been propagated to the Director through polling the Registry; skipping processing the downtimes in the server ads", serverName)
+			return
+		}
+	}
+
```

Comment:
This comment isn't for this exact snippet of code, but one of the things we talked about in our team discussion of downtime management was having Origins/Caches always report downtimes directly to the Director as well as through the Registry. We brought up the "CalDAV" protocol and mentioned that part of what the Director should be doing is figuring out how to merge two potentially-conflicting downtimes when one comes from the Origin/Cache and the other comes from the Registry.

Do you think we should take a stab at that in this PR so we can retain the functionality where downtimes are sent by Origins/Caches to the Director, or do you think that should be a followup to be tackled in the future?